### PR TITLE
code-cursor: init at 0.40.4

### LINKED ISSUE #309541: Package request: code-cursor
**Project description**
This is Code OSS fork with builtin AI code generator

**Metadata**

* homepage URL: https://cursor.sh/
* source URL: https://github.com/getcursor/cursor (I haven't idea how to get direct link from it)
* license: mit, unfree (I guess)
* platforms: linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by sarahec**
## Description of changes

Adds the [cursor](https://cursor.com) editor, a vscode fork with comprehensive AI integration for software development. (This is a FHS build via AppImage.)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux - n/a
  - [ ] x86_64-darwin - n/a
  - [ ] aarch64-darwin - n/a
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Closes https://github.com/NixOS/nixpkgs/issues/309541

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sarahec** [comment]
[Force pushes] "I have reformatted the file in the RFC style. Pray that I do not have to format it again."

File: pkgs/by-name/cu/cursor/package.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  curl,
+  yq,
+  common-updater-scripts,
+}:
+let
+  pname = "cursor";
+  version = "0.3.96";
+  appKey = "230313mzl4w4u92";
+  src = fetchurl {
+    url = "https://download.todesktop.com/230313mzl4w4u92/cursor-0.39.6-build-240819ih4ta2fye-x86_64.AppImage";
```
Comment (dotlambda) [inline pkgs/by-name/cu/cursor/package.nix:17]:
```suggestion
    url = "https://download.todesktop.com/${appKey}/cursor-0.39.6-build-240819ih4ta2fye-x86_64.AppImage";
```

File: pkgs/by-name/cu/cursor/package.nix [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/
+    cp -r bin $out/bin
+
+    mkdir -p $out/share/${pname}
```
Comment (dotlambda) [inline pkgs/by-name/cu/cursor/package.nix:36]:
```suggestion
    mkdir -p $out/share/cursor
```
#277994

File: pkgs/by-name/cu/cursor/package.nix [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    cp -a ${appimageContents}/locales $out/share/${pname}
+    cp -a ${appimageContents}/resources $out/share/${pname}
+    cp -a ${appimageContents}/usr/share/icons $out/share/
+    install -Dm 644 ${appimageContents}/${pname}.desktop -t $out/share/applications/
+
+    substituteInPlace $out/share/applications/${pname}.desktop --replace "AppRun" "${pname}"
```
Comment (dotlambda) [inline pkgs/by-name/cu/cursor/package.nix:42]:
```suggestion
    substituteInPlace $out/share/applications/cursor.desktop --replace-fail "AppRun" "cursor"
```

**Review by dotlambda — COMMENTED** [review]
`meta.mainProgram` and `meta.sourceProvenance` are missing

**Comment by sarahec** [comment]
@dotlambda thank you. Changes applied.

**Comment by sarahec** [comment]
And...fixed a dumb typo

File: pkgs/by-name/cu/cursor/package.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/
+    cp -r bin $out/bin
+
+    mkdir -p $out/share/cursor
+    cp -a ${appimageContents}/locales $out/share/${pname}
```
Comment (dotlambda) [inline pkgs/by-name/cu/cursor/package.nix:37]: Come on.
```suggestion
    cp -a ${appimageContents}/locales $out/share/cursor
```

Comment (sarahec) [reply pkgs/by-name/cu/cursor/package.nix:37]: Thought I'd gotten that one. Sorry.

**Comment by sarahec** [comment]
Sorry about the remaining ${pname}'s. Fixed them all.

File: pkgs/by-name/cu/cursor/package.nix [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  version = "0.3.96";
+  appKey = "230313mzl4w4u92";
+  src = fetchurl {
+    url = "https://download.todesktop.com/${appKey}/cursor-0.39.6-build-240819ih4ta2fye-x86_64.AppImage";
+    hash = "sha256-huZTzIZFAYtGRMhXGC+1sd0l2s5913QkWM+QjCtsEl0=";
+  };
+  appimage = appimageTools.wrapType2 { inherit version pname src; };
+  appimageContents = appimageTools.extractType2 { inherit version pname src; };
+in
+stdenvNoCC.mkDerivation rec {
```
Comment (dotlambda) [inline pkgs/by-name/cu/cursor/package.nix:23]:
```suggestion
stdenvNoCC.mkDerivation {
```

Comment (sarahec) [reply pkgs/by-name/cu/cursor/package.nix:23]: Done.

Comment (sarahec) [reply pkgs/by-name/cu/cursor/package.nix]: Done.

File: pkgs/by-name/cu/cursor/package.nix (hunk 4 above)
Comment (sarahec) [reply pkgs/by-name/cu/cursor/package.nix]: Done.

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cursor</li>
  </ul>
</details>

**Comment by sarahec** [comment]
@dotlambda that seems to be everything. Please let me know if there's more to do or if we need additional reviewers.

**Comment by sarahec** [comment]
@Supersandro2000 I'd love an extra pair of eyes on this as it's my first submitting based on an  AppImage

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cursor</li>
  </ul>
</details>

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

File: pkgs/by-name/co/code-cursor/package.nix [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+}:
+let
+  pname = "cursor";
+  version = "0.40.1";
+  appKey = "230313mzl4w4u92";
+  src = fetchurl {
+    url = "https://download.todesktop.com/${appKey}/cursor-0.40.1-build-2408245thnycuzj-x86_64.AppImage";
+    hash = "sha256-TPCJxrXFfatKDv5XrE8nqGFIE/KKi1aUZ/uUtIeIgxc=";
+  };
+  appimage = appimageTools.wrapType2 { inherit version pname src; };
```
Comment (SuperSandro2000) [inline pkgs/by-name/co/code-cursor/package.nix:20]:
```suggestion
```

File: pkgs/by-name/co/code-cursor/package.nix [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  appimageContents = appimageTools.extractType2 { inherit version pname src; };
+in
+stdenvNoCC.mkDerivation {
+  inherit pname version;
+
+  src = appimage;
```
Comment (SuperSandro2000) [inline pkgs/by-name/co/code-cursor/package.nix:26]:
```suggestion
  src = appimageTools.wrapType2 { inherit version pname src; };
```

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise LGTM

File: pkgs/by-name/co/code-cursor/package.nix (hunk 6 above)
Comment (sarahec) [reply pkgs/by-name/co/code-cursor/package.nix]: Done.

File: pkgs/by-name/co/code-cursor/package.nix (hunk 7 above)
Comment (sarahec) [reply pkgs/by-name/co/code-cursor/package.nix]: Done.

**Comment by sarahec** [comment]
> otherwise LGTM

@SuperSandro2000 thank you. I applied your suggestions.

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

File: pkgs/by-name/co/code-cursor/package.nix [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  appimageTools,
+  makeWrapper,
+  writeShellApplication,
+  curl,
+  yq,
+  common-updater-scripts,
+}:
+let
+  pname = "cursor";
+  version = "0.40.1";
```
Comment (SuperSandro2000) [inline pkgs/by-name/co/code-cursor/package.nix:14]:
```suggestion
  version = "0.40.3";
```

Comment (sarahec) [reply pkgs/by-name/co/code-cursor/package.nix:14]: Ugh. ${BADWORD}

Fixed

**Comment by sarahec** [comment]
@SuperSandro2000 Fixed the lagging version number. Thanks for spotting that.

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

**Review by bobvanderlinden — APPROVED** [review]
I tried to run it against some projects on NixOS. Seems to work fine :+1:

**Comment by poyhen** [comment]
is a fhs variant planned for this? like [vscode-fhs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/editors/vscode/generic.nix#L107)

**Comment by jottr** [comment]
Can this please be merged now?

**Comment by sarahec** [comment]
> is a fhs variant planned for this? like [vscode-fhs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/editors/vscode/generic.nix#L107)

It is a FHS build (which is why it's not being specified as working on Darwin.)

**Comment by sarahec** [comment]
> Can this please be merged now?

It's not my call. I've filed comments in the right place to expedite merging.

**Comment by jottr** [comment]
I'm aware that it's not your call. :)
Thanks for putting in the work to get this merged.

**Comment by AlephByte** [comment]
## Bug Report: Crash on Info Popup Display in Settings

### Issue Description
The application crashes when certain info popups are displayed, particularly within the Settings tab.

### Steps to Reproduce
1. Navigate to Settings > Cursor Settings > Features
2. Locate "Always keep composer in bound" option
3. Hover over the toggleable button
4. Observe the On/Off text popup
5. Application crashes

### Additional Observations
- Similar crashes occur with almost any popup within the Settings tab
- The issue seems isolated to the "Editor" zone
- No crashes observed when hovering over files in the File Explorer sidebar

### Environment
- Window Manager: Hyprland

If needed I can provide more info about my system.

**Comment by jottr** [comment]
Did you check if this also happens on any wm that is not Hyprland?

**Comment by AlephByte** [comment]
> Did you check if this also happens on any wm that is not Hyprland?

No but I could try it later today.
Does it work correctly for you?

**Comment by sarahec** [comment]
> ## Bug Report: Crash on Info Popup Display in Settings.

Please post application bugs to cursor's [forum](https://forum.cursor.com/). They are out of scope for the nix derivation.

**Comment by sarahec** [comment]
@SuperSandro2000 Kindly add this to your merge queue.

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

**Comment by pective** [comment]
any update on when this might be merged?

**Comment by sarahec** [comment]
> any update on when this might be merged?

I bumped it on the "ready to merge" list. PRs with a large number of comments may get passed over for easier, cleaner-looking ones.

**Comment by bobvanderlinden** [comment]
Result of `nixpkgs-review pr 337331` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-cursor</li>
  </ul>
</details>

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**